### PR TITLE
Enable widget filtering in Interface Customization dialog

### PR DESCRIPTION
**Pull request by DelazJ**
using both widget name and label
image

Code is inspired from [qgsoptionsdialoghighlightwidgetsimpl](https://github.com/qgis/QGIS/blob/6749cb3d1fcc61cb0b13444badfa5bf56378129b/src/gui/qgsoptionsdialoghighlightwidgetsimpl.cpp#L219) and TBH I don't know whether the protected variables are well handled but it builds, helped me do what I needed it for with no crash. And here it's for your review.
 
## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/app/qgscustomization.cpp [hunk 1]
```diff
@@ -115,6 +116,46 @@ QTreeWidgetItem *QgsCustomizationDialog::item( const QString &path, QTreeWidgetI
...
+    item->setHidden( !visible );
+  };
+  setChildrenVisible( treeWidget->invisibleRootItem(), false );
+
+  QList<QTreeWidgetItem *> items = treeWidget->findItems( text, Qt::MatchContains | Qt::MatchRecursive, 0 );
+  items.append( treeWidget->findItems( text, Qt::MatchContains | Qt::MatchRecursive, 1 ) );
+  success = !items.empty();
+  mTreeInitialExpand.clear();
+
+  for ( QTreeWidgetItem *item : items )
```
Comment (nyalldawson) [inline src/app/qgscustomization.cpp]:
```suggestion
  for ( QTreeWidgetItem *item : qgis::as_const( items ) )
```

File: src/app/qgscustomization.h [hunk 2]
```diff
@@ -101,6 +102,10 @@ class APP_EXPORT QgsCustomizationDialog : public QMainWindow, private Ui::QgsCus
 
     QString mLastDirSettingsName;
     QSettings *mSettings = nullptr;
+
+  protected:
+    QMap<QTreeWidgetItem *, bool> mTreeInitialExpand = QMap<QTreeWidgetItem *, bool>();
```
Comment (nyalldawson) [inline src/app/qgscustomization.h]:
```suggestion
    QMap<QTreeWidgetItem *, bool> mTreeInitialExpand;
```

File: src/app/qgscustomization.h [hunk 3]
```diff
@@ -101,6 +102,10 @@ class APP_EXPORT QgsCustomizationDialog : public QMainWindow, private Ui::QgsCus
...
+    QMap<QTreeWidgetItem *, bool> mTreeInitialVisible = QMap<QTreeWidgetItem *, bool>();
```
Comment (nyalldawson) [inline src/app/qgscustomization.h]:
```suggestion
    QMap<QTreeWidgetItem *, bool> mTreeInitialVisible;
```

File: src/app/qgscustomization.cpp [hunk 4]
```diff
@@ -55,7 +55,8 @@ QgsCustomizationDialog::QgsCustomizationDialog( QWidget * parent, QSettings * se
   connect( actionExpandAll, &QAction::triggered, this, &QgsCustomizationDialog::actionExpandAll_triggered );
   connect( actionCollapseAll, &QAction::triggered, this, &QgsCustomizationDialog::actionCollapseAll_triggered );
   connect( actionSelectAll, &QAction::triggered, this, &QgsCustomizationDialog::actionSelectAll_triggered );
-  connect( mCustomizationEnabledCheckBox, &QCheckBox::toggled, this, &QgsCustomizationDialog::mCustomizationEnabledCheckBox_toggled );
+  connect( mCustomizationEnabledCheckBox, &QCheckBox::toggled, this, &QgsCustomizationDialog::enableCustomization );
+  connect( mLeFilter, &QgsFilterLineEdit::textChanged, this, &QgsCustomizationDialog::filterItems );
```
Comment (nyalldawson) [inline src/app/qgscustomization.cpp]:
```suggestion
  connect( mLeFilter, &QgsFilterLineEdit::textChanged, this, &QgsCustomizationDialog::filterItems );

  mLeFilter->setShowSearchIcon( true );
```

**Comment by nyalldawson** [comment]
Looks good!

I feel like you just level-uped in ability. :tada: !!

**Comment by DelazJ** [comment]
Mostly a copy paste but thanks. Having spent hours on trying to find the right code before I remember to check the options dialog's, this is anyway of great learning.

**Pull request merged**